relay-redis/src/real.rs

```diff
@@ -1,146 +1,209 @@
+    ///
+    /// Returns a new [`AsyncRedisConnection`] that can be used to execute Redis commands.
+    /// The connection is automatically returned to the pool when dropped.
+    pub async fn get_connection(&self) -> Result<AsyncRedisConnection, RedisError> {
+        let connection = match self {
+            Self::Cluster(pool) => {
+                AsyncRedisConnection::Cluster(pool.get().await.map_err(RedisError::Pool)?)
+            }
+            Self::Single(pool) => {
+                AsyncRedisConnection::Single(pool.get().await.map_err(RedisError::Pool)?)
+            }
+        };
+
+        Ok(connection)
     }
 
-    /// Return [`RedisClientStats`] for [`AsyncRedisClient`].
+    /// Returns statistics about the current state of the connection pool.
     ///
-    /// It will always return 0 for `idle_connections` and 1 for `connections` since we
-    /// are re-using the same connection.
+    /// Provides information about the number of active and idle connections in the pool,
+    /// which can be useful for monitoring and debugging purposes.
     pub fn stats(&self) -> RedisClientStats {
+        let status = match self {
+            Self::Cluster(pool) => pool.status(),
+            Self::Single(pool) => pool.status(),
+        };
+
         RedisClientStats {
-            idle_connections: 0,
-            connections: 1,
+            idle_connections: status.available as u32,
+            connections: status.size as u32,
+            max_connections: status.max_size as u32,
+            waiting_for_connection: status.waiting as u32,
         }
     }
-}
 
-/// A wrapper type for async redis connections.
-#[derive(Clone)]
-pub enum AsyncRedisConnection {
-    /// Variant for [`ClusterConnection`].
-    Cluster(ClusterConnection),
-    /// Variant for [`ConnectionManager`] using [`redis::aio::MultiplexedConnection`].
-    Single(ConnectionManager),
-}
-
-impl AsyncRedisConnection {
-    /// Creates an [`AsyncRedisConnection`] in cluster mode.
-    pub async fn cluster<'a>(
-        servers: impl IntoIterator<Item = &'a str>,
+    /// Builds a [`Pool`] given a type implementing [`Manager`] and [`RedisConfigOptions`].
+    fn build_pool<M: Manager, W: From<Object<M>>>(
+        manager: M,
         opts: &RedisConfigOptions,
-    ) -> Result<Self, RedisError> {
-        // connection timeout is set in `base_pool_builder` on the pool level
-        let client = ClusterClientBuilder::new(servers)
-            .response_timeout(Duration::from_secs(opts.read_timeout))
-            .connection_timeout(Duration::from_secs(opts.connection_timeout))
+    ) -> Result<Pool<M, W>, BuildError> {
+        Pool::builder(manager)
+            .max_size(opts.max_connections as usize)
+            .wait_timeout(Some(Duration::from_secs(opts.wait_timeout)))
+            .create_timeout(Some(Duration::from_secs(opts.create_timeout)))
+            .recycle_timeout(Some(Duration::from_secs(opts.recycle_timeout)))
```

Comment:
Should these timeouts be optional, or how do we represent no timeout?

Reply:
I implemented this in a way that forces timeouts because I think they are good practice, having unbounded work is not a good idea.

Reply:
I am not sure about the wait timeout, Relay currently does not deal well with a timedout Redis operation, maybe waiting is better than starting to drop all data, but I am torn myself on this.

Reply:
Given the spooler, our preferred way of dealing with problems is to keep data in memory as much as we can, especially if failures are expected to eventually recover.

Reply:
Exactly, with a timeout we will just discard the data instead of spooling.